### DISCUSSION TITLE
Dynamic relationship `append` same relationship several times

### DISCUSSION THREAD

**Discussion by hegeds** [sqlalchemy/sqlalchemy#8150] [Usage Questions] [answered]
Hi,

I have a workflow, which in my opinion produces an unexpected outcome.

I have two tables connected with a many-to-many relationship. The relationship is set up with `dynamic` lazy loading (see example):

```python
from sqlalchemy import create_engine, Integer, ForeignKey, Column, Table
from sqlalchemy.orm import declarative_base, relationship, Session

Base = declarative_base()

a_b_relationships = Table(
    "a_b_relationships",
    Base.metadata,
    Column("a_id", ForeignKey("a.id")),
    Column("b_id", ForeignKey("b.id")),
)

class A(Base):
    __tablename__ = "a"
    id = Column(Integer, primary_key=True)
    bs_in_a = relationship(
        "B",
        lazy="dynamic",
        secondary=a_b_relationships,
        backref="as_in_b",
    )


class B(Base):
    __tablename__ = "b"
    id = Column(Integer, primary_key=True)
```

When I append one instance of `B` to an instance of `A` several times it creates multiple rows in the database. When I query the relationship it will only load the instance of b in the results one time, not as many time as it exists in the table. But when I call the `count` method the result will be the actual number of times the relationship has been created. I find this problematic as the result from `query.count` will differ from the result `len(query.all()`

```python
with Session(engine) as session:
    a = A(id=1)
    b = B(id=2)

    session.add_all([a, b])
    session.commit()

    # b added to a twice
    a.bs_in_a.append(b)
    session.commit()

    a.bs_in_a.append(b)
    session.commit()

    count_query = "SELECT COUNT('b_id') FROM a_b_relationships"

    print("b in a:", a.bs_in_a.count())
    print("Relationships:", engine.execute(count_query).first()[0])

    assert a.bs_in_a.count() == engine.execute(count_query).first()[0]
    assert a.bs_in_a.count() == a.bs_in_a.all()
``` 

The other thing, is that when I try to remove the relationship with  `remove` , then it will throw the following error:
```
sqlalchemy.orm.exc.StaleDataError: DELETE statement on table 'a_b_relationships' expected to delete 1 row(s); Only 2 were matched.
```

**TL:DR**
My issue, is that during `append` the relationships seems to operate on the expectation  that it could have multiple instances of the same relationship, but the `remove` and `all` statement behaves, as there could only be one. Is this an expected behavior? If yes how can I create a guard which will disallow creating the same relationship multiple time? Is there a recommended way?

the whole example script:
```python
from sqlalchemy import create_engine, Integer, ForeignKey, Column, Table
from sqlalchemy.orm import declarative_base, relationship, Session

Base = declarative_base()

a_b_relationships = Table(
    "a_b_relationships",
    Base.metadata,
    Column("a_id", ForeignKey("a.id")),
    Column("b_id", ForeignKey("b.id")),
)

class A(Base):
    __tablename__ = "a"
    id = Column(Integer, primary_key=True)
    bs_in_a = relationship(
        "B",
        lazy="dynamic",
        secondary=a_b_relationships,
        backref="as_in_b",
    )


class B(Base):
    __tablename__ = "b"
    id = Column(Integer, primary_key=True)


engine = create_engine("mysql://root:root@127.0.0.1/test")

Base.metadata.drop_all(engine)
Base.metadata.create_all(engine)


with Session(engine) as session:
    a = A(id=1)
    b = B(id=2)

    session.add_all([a, b])
    session.commit()

    # b added to a twice
    a.bs_in_a.append(b)
    session.commit()

    a.bs_in_a.append(b)
    session.commit()

    count_query = "SELECT COUNT('b_id') FROM a_b_relationships"

    print("b in a:", a.bs_in_a.count())
    print("Relationships:", engine.execute(count_query).first()[0])
    print("Length of `bs_in_a`:", len(a.bs_in_a.all()))

    assert a.bs_in_a.count() == engine.execute(count_query).first()[0]
    assert a.bs_in_a.count() == a.bs_in_a.all()
```

**Comment by zzzeek** [comment, marked as answer]
so everything here is expected.   it seems the most problematic part is that you see the discrepancy between count() and query.all(), which is a problem, and the solution for that part of it is that [query is now legacy](https://docs.sqlalchemy.org/en/14/changelog/migration_20.html#orm-query-unified-with-core-select) and the implicit "deduping" behavior [does not apply to 2.0 style executions](https://docs.sqlalchemy.org/en/14/changelog/migration_20.html#orm-rows-not-uniquified-by-default), so you will not see that discrepancy when using new APIs.

however when using "dynamic", we've not fully created what the new API will be since the way "dynamic" works is already not very 2.0-ish.   issue #7123 contains where we will decide what the best pattern will be, however you can use dynamic relationships right now using `.statement`.  Rewriting your code using 2.0 style execution, all assertions pass.

```py
    count_query = text("SELECT COUNT('b_id') FROM a_b_relationships")

    print(
        "b in a:",
        session.scalar(
            select(func.count("*")).select_from(a.bs_in_a.subquery())
        ),
    )
    print("Relationships:", session.execute(count_query).first()[0])
    print(
        "Length of `bs_in_a`:", len(session.scalars(a.bs_in_a.statement).all())
    )

    assert (
        session.scalar(
            select(func.count("*")).select_from(a.bs_in_a.subquery())
        )
        == session.execute(count_query).first()[0]
    )
    assert (
        session.scalar(
            select(func.count("*")).select_from(a.bs_in_a.subquery())
        )
        == len(session.scalars(a.bs_in_a.statement).all())
    )

```

as far as append() and remove() not being symmetric, this is a product of how SQL works; we can INSERT as many duplicate rows as we want in SQL, however we cannot DELETE just one of them if the rows are fully identical.      The usual solution is that your `a_b_relationships` table should not allow duplicates, as seen in the [docs for many to many](https://docs.sqlalchemy.org/en/14/orm/basic_relationships.html#many-to-many):

```py
a_b_relationships = Table(
    "a_b_relationships",
    Base.metadata,
    Column("a_id", ForeignKey("a.id"), primary_key=True),
    Column("b_id", ForeignKey("b.id"), primary_key=True),
)
```

**Reply by hegeds** [reply]
Thank you for the detailed anserw, I think I understand it now!
